python/hail/expr.py

```diff
@@ -39,6 +41,9 @@ def __eq__(self, other):
     def __hash__(self):
         return self._jtype.hashCode()
 
+    def _set_history(self, history):
```

Comment:
I think you should have a `HasHistory` mixin.  Then in record, you can check if an object is an instance of `HasHistory` rather than it having a history `_history` attribute (which is a bit dangerous).